app/src/main/java/com/babylon/wallet/android/presentation/accessfactorsources/signatures/GetSignaturesViewModel.kt

```diff
@@ -0,0 +1,261 @@
+                signers = signers,
+                signRequest = signRequest
+            ).onSuccess { signatures ->
+                signaturesWithPublicKeys.addAll(signatures)
+            }.onFailure {
+                accessFactorSourcesUiProxy.setOutput(
+                    AccessFactorSourcesOutput.Failure(error = it)
+                )
+            }.also {
+                isSigningWithDeviceInProgress.emit(false)
+            }
+        }
+    }
+
+    private suspend fun collectSignaturesForLedgerFactorSource(
+        ledgerFactorSource: FactorSource.Ledger,
+        signers: List<ProfileEntity>,
+        signRequest: SignRequest
+    ) {
+        collectSignaturesWithLedgerJob?.cancel()
+        collectSignaturesWithLedgerJob = viewModelScope.launch {
+            signWithLedgerFactorSourceUseCase(
+                ledgerFactorSource = ledgerFactorSource,
+                signers = signers,
+                signRequest = signRequest
+            ).onSuccess { signatures ->
+                signaturesWithPublicKeys.addAll(signatures)
+            }.onFailure {
+                accessFactorSourcesUiProxy.setOutput(
+                    AccessFactorSourcesOutput.Failure(error = it)
+                )
+            }.also {
+                isSigningWithLedgerInProgress.emit(false)
+            }
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private suspend fun getSigningEntitiesByFactorSource(
+        signers: List<ProfileEntity>
+    ): Map<FactorSource, List<ProfileEntity>> = withContext(defaultDispatcher) {
+        val result = mutableMapOf<FactorSource, List<ProfileEntity>>()
+        val profile = getProfileUseCase()
+
+        signers.forEach { signer ->
+            when (val securityState = signer.securityState) {
+                is EntitySecurityState.Unsecured -> {
+                    val factorSourceId = securityState.value.transactionSigning.factorSourceId.asGeneral()
+                    val factorSource = requireNotNull(profile.factorSourceById(factorSourceId))
+
+                    if (factorSource.kind != FactorSourceKind.TRUSTED_CONTACT) { // trusted contact cannot sign!
+                        if (result[factorSource] != null) {
+                            result[factorSource] = result[factorSource].orEmpty() + listOf(signer)
+                        } else {
+                            result[factorSource] = listOf(signer)
+                        }
+                    }
+                }
+            }
+        }
+
+        result.keys
+            .sortedBy { factorSource ->
+                factorSource.kind.signingOrder()
+            }.associateWith { factorSource ->
+                result[factorSource]!!
+            }
+    }
+
+    private fun FactorSourceKind.signingOrder(): Int { // based on difficulty - most "challenging" factor comes first
+        return when (this) {
+            FactorSourceKind.LEDGER_HQ_HARDWARE_WALLET -> 0
```

Comment:
We had device first before as we asked once for biometrics for whole signing process, so we wanted first to collect device signatures while we have access to keystore, as collecting ledger signatures might take long enough for the keystore access window to expire. I see that you switched the order, so I assume that you also changed how we ask for biometrics?

Reply:
yes. We now ask the most challenging factor sources to sign and device should be the last. 
When it's time for device factor source user gets the biometric prompt